### DISCUSSION TITLE
ignore

### DISCUSSION THREAD

**Discussion by Mr-Programmer-777** [ScratchAddons/ScratchAddons#1817] [Discussion]
Okeydokes, is it ok with you guys if you give me a pop quiz kinda thing?

**Comment by Mr-Programmer-777** [comment]
Also, can ATOM be used inside GitHub or...?

**Reply by cobaltt7** [reply]
Atom is an app that can be used to write code. I use Visual Studio Code. [Soon VS code will be a web editor for github](https://github.com/features/codespaces)

**Reply by Mr-Programmer-777** [reply]
Right. So if i was to do `getElementById()` i could get a element, and change it? (I'm at that bit now)

**Reply by cobaltt7** [reply]
yes, you could `document.getElementById(id)` for that if you know the id from the HTML.


FYI, `innerHTML` and `insertAdjacentHTML` aren't allowed, see #173

**Reply by Mr-Programmer-777** [reply]
I think I get it now, so using a userscript, you can `getElementById`, change the HTML and CSS, and call it a new theme?

**Reply by cobaltt7** [reply]
well, for a theme, you don't need any JS. just use a css file

**Reply by Mr-Programmer-777** [reply]
Cool. I have to go to bed now, so ill see you guys in... Around 16 hours.
Quick question, how would you change current CSS using your own userstyle?

**Reply by cobaltt7** [reply]
say scratch's css has this:
```css
color: white;
```
and you want to change it to black. just do
```css
color: black;
```
in your css and it will be overridden

notes:

- if you have the `theme` tag in your addon manifest, the css will be inputted into a `<style>` element. that will override all external styesheets, but not other `<style>` elements.

- here is a grid. if your selector is further down the list then scratch's, it will override scratch's.
![](https://specifishity.com/specifishity.png)

- if you have to, you can add `!important` to the css like this

```css
color: black!important;
```
to override scratch

**Reply by Mr-Programmer-777** [reply]
Hey! I'm back. So, you can use `#document`, find a `div` Class or ID, and use something like:
`div id = banner {
 color = #ffffff;
}`
To change the blue to white?

**Reply by cobaltt7** [reply]
> Hey! I'm back. So, you can use `#document`, find a `div` Class or ID, and use something like:
> `div id = banner { color = #ffffff; }`
> To change the blue to white?

```css
div#banner {
    color: #ffffff;
}
```

**Reply by cobaltt7** [reply]
here is a css selector video https://www.youtube.com/watch?v=l1mER1bV0N0

**Reply by mxmou** [reply]
> `div #banner { color = #ffffff; }` or `div [id=banner] { color = #ffffff; }`

The `#banner` syntax is preferred because it has higher specificity. And the equal sign should be a colon.

**Reply by cobaltt7** [reply]
> The `#banner` syntax is preferred because it has higher specificity.

yeah


> And the equal sign should be a colon.

oop

**Reply by Mr-Programmer-777** [reply]
The syntax is probably wrong but uh oops...

**Reply by Mr-Programmer-777** [reply]
Aha! Ok, I'll keep all that in mind. I'm gonna try to make a new theme.

**Reply by Mr-Programmer-777** [reply]
Should I first disable everything, so it doesn't mess it up?

**Reply by cobaltt7** [reply]
> Should I first disable everything, so it doesn't mess it up?

your choice. it should work well if any other addons are enabled, but also if no others.

**Reply by Mr-Programmer-777** [reply]
Uhh, it says I HAVE to have Chrome...
I have firefox...

**Reply by cobaltt7** [reply]
> Uhh, it says I HAVE to have Chrome...
> I have firefox...

what says

**Reply by Mr-Programmer-777** [reply]
https://github.com/ScratchAddons/ScratchAddons/wiki/Creating-an-addon

**Reply by mxmou** [reply]
> https://github.com/ScratchAddons/ScratchAddons/wiki/Creating-an-addon

Chrome is recommended because it's easier to work with, but everything works in Firefox too.

**Reply by Mr-Programmer-777** [reply]
Phew.

**Reply by Mr-Programmer-777** [reply]
Do you know if there is a ID or Class that selects the whole header of the editor?

**Reply by mxmou** [reply]
> Do you know if there is a ID or Class that selects the whole header of the editor?

`[class^=menu-bar_main-menu]`

You can use right click → inspect to find the class name, but it's best to use an attribute selector in addons so that you don't need to include the hash (the part after the last underscore).

**Reply by Mr-Programmer-777** [reply]
Wait i think i got it

**Reply by Mr-Programmer-777** [reply]
Ok, thanks

**Reply by Mr-Programmer-777** [reply]
I'm having trouble compiling an edited .zip of Scratch Addons; Firefox keeps on spurting `File /home/username/Documents/Scratch Projects/scratchAddons builds/ScratchAddons-master_edited.zip does not contain a valid manifest` when i haven't even touched `manifest.json`... Why?

**Reply by mxmou** [reply]
> I'm having trouble compiling an edited .zip of Scratch Addons; Firefox keeps on spurting `File /home/username/Documents/Scratch Projects/scratchAddons builds/ScratchAddons-master_edited.zip does not contain a valid manifest` when i haven't even touched `manifest.json`... Why?

Try loading the manifest itself instead of a zip.

**Reply by Mr-Programmer-777** [reply]
Soo... Just `manifest.json`?

**Reply by Mr-Programmer-777** [reply]
I got this: `Extension is invalid; Reading manifest: Error processing incognito: Invalid enumeration value "not-allowed"`

**Reply by Mr-Programmer-777** [reply]
Maybe it's because it's Firefox, why don't I remove `incognito`?

**Reply by Mr-Programmer-777** [reply]
I'll try that, hold on

**Reply by Mr-Programmer-777** [reply]
It works!

**Reply by Mr-Programmer-777** [reply]
My new addon doesn't show up though..
I mean its not finished so i dont blame it
OOH wait that was just the manifest
oo

**Reply by Mr-Programmer-777** [reply]
UGHH. I've tried everything to get the vanilla ScratchAddons.zip to work but it's not having it, it keeps on saying it has a invalid manifest...

**Comment by cobaltt7** [comment]
unzip it

**Reply by Mr-Programmer-777** [reply]
I did, it's saying `Reading manifest: Error processing browser_specific_settings.gecko.id: Value "griffpatch@griffpatch.co.uk" must either: match the pattern /^\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\}$/i, or match the pattern /^[a-z0-9-._]*@[a-z0-9-._]+$/i`

**Reply by Mr-Programmer-777** [reply]
I did, it's saying `Reading manifest: Error processing browser_specific_settings.gecko.id: Value "griffpatch@griffpatch.co.uk" must either: match the pattern /^\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\}$/i, or match the pattern /^[a-z0-9-._]*@[a-z0-9-._]+$/i`

**Reply by Mr-Programmer-777** [reply]
I did, it's saying `Reading manifest: Error processing browser_specific_settings.gecko.id: Value "griffpatch@griffpatch.co.uk" must either: match the pattern /^\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\}$/i, or match the pattern /^[a-z0-9-._]*@[a-z0-9-._]+$/i`

**Reply by Mr-Programmer-777** [reply]
I did, it's saying `Reading manifest: Error processing browser_specific_settings.gecko.id: Value "griffpatch@griffpatch.co.uk" must either: match the pattern /^\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\}$/i, or match the pattern /^[a-z0-9-._]*@[a-z0-9-._]+$/i`

**Reply by Mr-Programmer-777** [reply]
OH GOD
SORRY
GITHUB WHAT THE HELL

**Reply by Mr-Programmer-777** [reply]
OH GOD
SORRY
GITHUB WHAT THE HELL

**Reply by Mr-Programmer-777** [reply]
AGH

**Reply by Mr-Programmer-777** [reply]
I changed that ID thing and now its saying crap: `Extension is invalid

Reading manifest: Error processing browser_specific_settings.gecko.id: Value "/^[a-z0-9-._]*@[a-z0-9-._]+$/i" must either: match the pattern /^\{[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\}$/i, or match the pattern /^[a-z0-9-._]*@[a-z0-9-._]+$/i`

**Reply by cobaltt7** [reply]
try `addon@example.com`

**Reply by Mr-Programmer-777** [reply]
Is that a email or..

**Reply by Mr-Programmer-777** [reply]
Now what?

**Reply by cobaltt7** [reply]
wait what firefox version do you have

**Reply by Mr-Programmer-777** [reply]
Mozilla Firefox (86.0)

**Reply by Mr-Programmer-777** [reply]
I read something about IDs on help but idk

**Reply by cobaltt7** [reply]
try removing the whole `browser_specific_settings` object

**Reply by Mr-Programmer-777** [reply]
Right, ok, ill do that

**Reply by Mr-Programmer-777** [reply]
JSON.parse: expected double-quoted property name at line 59 column 5 of the JSON data

**Reply by Mr-Programmer-777** [reply]
oops my bad

**Reply by cobaltt7** [reply]
you must have a trailing comma

**Reply by Mr-Programmer-777** [reply]
yeah i think i fixed it wait

**Reply by Mr-Programmer-777** [reply]
WOOO it's done it!

**Reply by Mr-Programmer-777** [reply]
It works!!
I'll compile the whole zip and do it

**Reply by cobaltt7** [reply]
great!

**Reply by Mr-Programmer-777** [reply]
Nope.
Still NoT a VaLiD MaNiFeSt

**Reply by Mr-Programmer-777** [reply]
Can you see if it works on Chrome?
[ScratchAddons-master_3.zip](https://github.com/ScratchAddons/ScratchAddons/files/6132366/ScratchAddons-master_3.zip)

**Reply by cobaltt7** [reply]
wait what


whats the error?

**Reply by Mr-Programmer-777** [reply]
Still `ScratchAddons-master_3.zip does not contain a valid manifest`

**Reply by Mr-Programmer-777** [reply]
I'll only add my addon when it works, if you were wondering

**Reply by mxmou** [reply]
> Nope.
> Still NoT a VaLiD MaNiFeSt

Try using the unchanged manifest, it should work. Make sure to disable any other Scratch Addons installation, and you don't need to zip it. Select the `manifest.json` file after clicking "Load temporary add-on".

**Reply by cobaltt7** [reply]
yeah it works...

**Reply by Mr-Programmer-777** [reply]
> > Nope.
> > Still NoT a VaLiD MaNiFeSt
> 
> Try using the unchanged manifest, it should work. Make sure to disable any other Scratch Addons installation, and you don't need to zip it. Select the `manifest.json` file after clicking "Load temporary add-on".

The complete raw, out-of-the-repo one?

**Reply by mxmou** [reply]
Oh, and the issue with the zip is that the extension directory is *inside* the zip, which is meant to contain the folder contents, not the folder itself.

**Reply by Mr-Programmer-777** [reply]
> Oh, and the issue with the zip is that the extension directory is _inside_ the zip, which is meant to contain the folder contents, not the folder itself.

Can you compile what you just said?

**Reply by mxmou** [reply]
> > > Nope.
> > > Still NoT a VaLiD MaNiFeSt
> > 
> > 
> > Try using the unchanged manifest, it should work. Make sure to disable any other Scratch Addons installation, and you don't need to zip it. Select the `manifest.json` file after clicking "Load temporary add-on".
> 
> The complete raw, out-of-the-repo one?

Yes. It shouldn't be necessary to change the manifest, just load it into the browser.

**Reply by Mr-Programmer-777** [reply]
So what if i try a zip with all the folders (like 'addons' or 'libs') emptied out?

**Reply by cobaltt7** [reply]
> So what if i try a zip with all the folders (like 'addons' or 'libs') emptied out?

that wont work

**Reply by mxmou** [reply]
> > Oh, and the issue with the zip is that the extension directory is _inside_ the zip, which is meant to contain the folder contents, not the folder itself.
> 
> Can you compile what you just said?

Currently the structure of your zip file is like this:
```
ScratchAddons-master.zip
  ScratchAddons-master
    manifest.json
    (other files)
```

Firefox expects this:
```
ScratchAddons-master.zip
  manifest.json
  (other files)
```

A possible solution would be to select files inside the ScratchAddons-master folder and not the folder itself when creating the zip, but the best solution is to not use a zip at all, just select the manifest file.

**Reply by Mr-Programmer-777** [reply]
Guys, I really appreciate your help on this, and I know I'm making you rage at your screens, but I try, ok?
Give me a minute, im gonna wing it.

**Reply by Mr-Programmer-777** [reply]
Right. I have progress (i think). The zip works, but I'm not sure a box saying Scratch Addons is good enough. Yeah, it's a box with no addons showing. brb

**Reply by Mr-Programmer-777** [reply]
Inspected it, the poor thing only has a `head`. Literally.

**Reply by mxmou** [reply]
> Right. I have progress (i think). The zip works, but I'm not sure a box saying Scratch Addons is good enough. Yeah, it's a box with no addons showing. brb

Make sure you don't have two Scratch Addons enabled at the same time.

**Reply by Mr-Programmer-777** [reply]
I don't, but I found this in the console: Unchecked lastError value: Error: The storage API will not work with a temporary addon ID. Please add an explicit addon ID to your manifest. For more information see https://mzl.la/3lPk1aE.

**Reply by Mr-Programmer-777** [reply]
Also, as you said, I zipped it inside the folder, not zip the folder.

**Reply by mxmou** [reply]
> I don't, but I found this in the console: Unchecked lastError value: Error: The storage API will not work with a temporary addon ID. Please add an explicit addon ID to your manifest. For more information see https://mzl.la/3lPk1aE.

Maybe the `browser_specific_settings` are missing in the manifest?

**Reply by cobaltt7** [reply]
We had to remove it earlier, it wasn't matching the regex

**Reply by Mr-Programmer-777** [reply]
what if i add that without the gecko thing or am i just mad

**Reply by Mr-Programmer-777** [reply]
(LAZY MODE ACTIVATE)

**Reply by Mr-Programmer-777** [reply]
Hold on: whats the point of making a fork of SA?
Can't i just code the addon and send it off?

**Reply by Mr-Programmer-777** [reply]
Plus, the manifest can load all files it's given, so in theory i can test it like a extension. if it works, i can sent it off.

**Reply by cobaltt7** [reply]
This is why I just use git...

**Reply by Mr-Programmer-777** [reply]
?? What's that? Is it like an online editor?

**Reply by cobaltt7** [reply]
it's an program that let's you easily get code from and put code on github

**Reply by cobaltt7** [reply]
Are you on windows or mac?

**Reply by Mr-Programmer-777** [reply]
Linux

**Reply by Mr-Programmer-777** [reply]
Also update: i give up.

**Reply by Mr-Programmer-777** [reply]
I really, REALLY wanted to make a rigid, nice theme that looked like Visual Studio, with all the editor add-ons, and a little box that said thanks for using my theme and.. It's just too hard.

**Comment by Lurux** [comment]
FYI, my usual rule of thumb is to limit myself to 4 open issues at once in any project.
If I want to add more I'll usually see if I can close an outdated issue, else I just wait for one of them to be triaged.
This helps to keep things clean on the maintainer's side, and it helps to concentrate on the most important things first :smile:

Edit: posting this after seeing the discussion in issue 1801
